README.txt
README - MMC3 code for neslib based C (cc65)
(version Feb 2023)

DEMO CONTROLS
Up/Down/Right/Left - move sprites
A or B - scroll the screen


 
In this configuration, A000-FFFF is fixed and only 8000-9FFF 
is swappable. Use the banked_call() system that cppchriscpp 
designed. If you are calling a function in a swappable bank, 
use a banked_call().

CHR (graphics) is set to "INVERTED", so that we can cycle through 
BG tiles using the smallest amount of tile banks.
You can change this (top of mmc3_code.asm)...A12_INVERT...to zero.

The A12_INVERT is a CONSTANT value in this code. I suppose you could 
make it variable...with many changes to the asm code.

;if invert bit is 0 - this is if you want tiny sprite banks
;mode 0 changes $0000-$07FF
;mode 1 changes $0800-$0FFF
;mode 2 changes $1000-$13FF
;mode 3 changes $1400-$17FF
;mode 4 changes $1800-$1BFF
;mode 5 changes $1C00-$1FFF

;if invert bit is $80 - this is if you want tiny BG banks
;mode 0 changes $1000-$17FF
;mode 1 changes $1800-$1FFF
;mode 2 changes $0000-$03FF
;mode 3 changes $0400-$07FF
;mode 4 changes $0800-$0BFF
;mode 5 changes $0C00-$0FFF

For the scanline IRQ to work properly, you must set BG to use set 
#0 and sprites to use #1. And the screen needs to be ON.






...

The scanline IRQ uses an automated system that I developed. It uses 
a char array that describes how the screen will set scanline counts, 
and gives you some options about what happens at that scanline.

IRQ SYSTEM FORMAT = 
value < 0xf0, it's a scanline count
zero is valid, it triggers an IRQ at the end of the current line

if >= 0xf0...
f0 = 2000 write, next byte is write value
f1 = 2001 write, next byte is write value
f2-f4 unused - future TODO ?
f5 = 2005 write, next byte is H Scroll value
f6 = 2006 write, next 2 bytes are write values

f7 = change CHR mode 0, next byte is write value
f8 = change CHR mode 1, next byte is write value
f9 = change CHR mode 2, next byte is write value
fa = change CHR mode 3, next byte is write value
fb = change CHR mode 4, next byte is write value
fc = change CHR mode 5, next byte is write value

fd = very short wait, no following byte 
fe = short wait, next byte is quick loop value
(for fine tuning timing of things)

ff = end of data set
Once it sees a scanline value or an 0xff, it exits the parser.

So, if you want a CHR swap to affect the entire screen, put it first,
then the first IRQ scanline count (or 0xff end).
The next thing after a scanline count won't happen until the next
IRQ fires. You can then change scroll, or whatever.

Small example...
irq_array[0] = 47; // wait 48 lines
...
irq_array[1] = 0xf5; // 2005, H scroll change
irq_array[2] = 0; // change it to zero
irq_array[3] = 0xff; // end of data

Note: the very top of the screen will use the H scroll value passed to
the neslib scroll() function (or my set_scroll_x function).

Passing the array address to set_irq_ptr() turns on the system. Try
not to extend your game logic past 1 frame and only have the array
half written... something bad could happen. The system can't swap
PRG banks, so, it shouldn't crash the game.

Note: use extreme caution writing to the $2000 register. I suppose you
could change the nametable selection, but don't change the other bits
or something bad will happen.

Note that anything above the first scanline count (in the mmc3
IRQ array) will happen inside the NMI code, before the next frame
starts drawing to the screen.



Warnings:

Changes have been made, Feb 2023, if interrupts are writing 
values to $8000, the previous $8000 setting will be restored 
(to prevent the wrong bank swapped). 
It's safer and less picky (and the main code should be able 
to do anything you like now).

The system works as long as the IRQ system doesn't get
interrupted by NMI, while trying to to a CHR bank change.

Don't tell the IRQ system to change a CHR bank near the bottom
of the screen. If you need to change the CHR bank between
frames (so it applies at the top of the next frame) you should
put that at the very top of the IRQ array, before the
scanline count.

The new changes might effect the timing of IRQ events.




Other functions
---------------
set_prg_8000() change the swapped bank
get_prg_8000() get the bank #
set_mirroring() change the mirroring
disable_irq() turn off the irq system
set_irq_ptr() pass the address of an array to the system, turn it on

set_chr_mode_0() change a chr bank
set_chr_mode_1() etc
set_chr_mode_2()
set_chr_mode_3()
set_chr_mode_4()
set_chr_mode_5()


I wouldn't recommend using these functions, but they are available...

set_prg_a000() !! could crash with the current cfg
set_wram_mode() you shouldn't need to touch this


Just to repeat. To call a function in a swapped bank, use banked_call().
It will automatically swap the correct bank in, and call the function.
If you just need to access data stored in a swappable bank, then you
can call set_prg_8000() to put it in place, first, before reading the 
data.

To swap CHR banks, either use the irq system, or call one of the CHR
functions, but make sure the irq system isn't also changing CHR banks.



IRQ SYSTEM WAIT TIMINGS
 
fd short wait 1/10 of scanline (36 pixels)
fe, wait 0 = 100 pixels, or 1/3 scanline
fe, wait 5 = 1/2 scanline
fe, wait 16 = 1 scanline
fe, wait 39 ($27) = 2 scanlines
wait x scanlines = ((x-1)*23) + 15
fe, wait 255 ($ff) = 11 1/2 scanlines

These could be used to better time writes to 2005 or 2006


A hardware note
$8000 register = CP-- -RRR	
; P = 0, set PRG $8000 swappable and $c000 fixed to the next to last bank
; P = 1, set PRG $c000 swappable and $8000 fixed to the next to last bank
; C = 0, set CHR two 2 KB banks at $0000-$0FFF, four 1 KB banks at $1000-$1FFF
; C = 1, set CHR two 2 KB banks at $1000-$1FFF, four 1 KB banks at $0000-$0FFF
; C called "A12_INVERT" at top of mmc3_code.asm
; RRR = mode

MODES
0: Select 2 KB CHR bank at PPU $0000-$07FF (or $1000-$17FF)*
1: Select 2 KB CHR bank at PPU $0800-$0FFF (or $1800-$1FFF)*
2: Select 1 KB CHR bank at PPU $1000-$13FF (or $0000-$03FF)*
3: Select 1 KB CHR bank at PPU $1400-$17FF (or $0400-$07FF)*
4: Select 1 KB CHR bank at PPU $1800-$1BFF (or $0800-$0BFF)*
5: Select 1 KB CHR bank at PPU $1C00-$1FFF (or $0C00-$0FFF)*

6: Select 8 KB PRG ROM bank at $8000-$9FFF (or $C000-$DFFF)**
7: Select 8 KB PRG ROM bank at $A000-$BFFF

*depending on C value "A12_INVERT"
**depending on P value, the other will be fixed to the next to last bank



Another warning:

I put a double buffering in, to make sure that we aren't
changing the array while the IRQ is trying to read it.
...but it could have bugs.
If you the frame (main code) runs long, it could
be interrupted by NMI while copying, and cause problems.

TODO - improve/fix that.




CHANGES.txt
CHANGES
10/2/2022

last pop wasn't setting the default
PRG bank back into place. Fixed.

Feb 2023
the "safe" bank changes were not
thread safe. That has been fixed.

PAL_BUF has been moved out of the stack
and into the BSS segment.



crt0.s
; Startup code for cc65 and Shiru's NES library
; based on code by Groepaz/Hitmen <groepaz@gmx.net>, Ullrich von Bassewitz <uz@cc65.org>




; Edited to work with MMC3 code
.define SOUND_BANK 12
;segment BANK12


FT_BASE_ADR		= $0100		;page in RAM, should be $xx00
FT_DPCM_OFF		= $f000		;$c000..$ffc0, 64-byte steps
FT_SFX_STREAMS	= 1			;number of sound effects played at once, 1..4

FT_THREAD       = 1		;undefine if you call sound effects in the same thread as sound update
FT_PAL_SUPPORT	= 1		;undefine to exclude PAL support
FT_NTSC_SUPPORT	= 1		;undefine to exclude NTSC support
FT_DPCM_ENABLE  = 1		;undefine to exclude all DMC code
FT_SFX_ENABLE   = 1		;undefine to exclude all sound effects code





;REMOVED initlib
;this called the CONDES function

    .export _exit,__STARTUP__:absolute=1
	.import push0,popa,popax,_main,zerobss,copydata

; Linker generated symbols
	.import __STACK_START__   ,__STACKSIZE__ ;changed
	.import __ROM0_START__  ,__ROM0_SIZE__
	.import __STARTUP_LOAD__,__STARTUP_RUN__,__STARTUP_SIZE__
	.import	__CODE_LOAD__   ,__CODE_RUN__   ,__CODE_SIZE__
	.import	__RODATA_LOAD__ ,__RODATA_RUN__ ,__RODATA_SIZE__
	.import NES_MAPPER, NES_PRG_BANKS, NES_CHR_BANKS, NES_MIRRORING

    .include "zeropage.inc"




PPU_CTRL	=$2000
PPU_MASK	=$2001
PPU_STATUS	=$2002
PPU_OAM_ADDR=$2003
PPU_OAM_DATA=$2004
PPU_SCROLL	=$2005
PPU_ADDR	=$2006
PPU_DATA	=$2007
PPU_OAM_DMA	=$4014
PPU_FRAMECNT=$4017
DMC_FREQ	=$4010
CTRL_PORT1	=$4016
CTRL_PORT2	=$4017

OAM_BUF		=$0200
;PAL_BUF		=$01c0
VRAM_BUF	=$0700


.segment "BSS"
PAL_BUF: .res 32
;move this out of the hardware stack
;the mmc3 code is using more of the stack
;and might collide with $1c0-1df



.segment "ZEROPAGE"

NTSC_MODE: 			.res 1
FRAME_CNT1: 		.res 1
FRAME_CNT2: 		.res 1
VRAM_UPDATE: 		.res 1
NAME_UPD_ADR: 		.res 2
NAME_UPD_ENABLE: 	.res 1
PAL_UPDATE: 		.res 1
PAL_BG_PTR: 		.res 2
PAL_SPR_PTR: 		.res 2
SCROLL_X: 			.res 1
SCROLL_Y: 			.res 1
SCROLL_X1: 			.res 1
SCROLL_Y1: 			.res 1
PAD_STATE: 			.res 2		;one byte per controller
PAD_STATEP: 		.res 2
PAD_STATET: 		.res 2
PPU_CTRL_VAR: 		.res 1
PPU_CTRL_VAR1: 		.res 1
PPU_MASK_VAR: 		.res 1
RAND_SEED: 			.res 2
FT_TEMP: 			.res 3

TEMP: 				.res 11
SPRID:				.res 1

PAD_BUF		=TEMP+1

PTR			=TEMP	;word
LEN			=TEMP+2	;word
NEXTSPR		=TEMP+4
SCRX		=TEMP+5
SCRY		=TEMP+6
SRC			=TEMP+7	;word
DST			=TEMP+9	;word

RLE_LOW		=TEMP
RLE_HIGH	=TEMP+1
RLE_TAG		=TEMP+2
RLE_BYTE	=TEMP+3

;nesdoug code requires
VRAM_INDEX:			.res 1
META_PTR:			.res 2
DATA_PTR:			.res 2





.segment "HEADER"

    .byte $4e,$45,$53,$1a
	.byte <NES_PRG_BANKS
	.byte <NES_CHR_BANKS
	.byte <NES_MIRRORING|(<NES_MAPPER<<4)
	.byte <NES_MAPPER&$f0
	.byte 1 ;8k of PRG RAM
	.res 7,0


; linker complains if I don't have at least one mention of each bank
.segment "ONCE"
.segment "BANK0"
.segment "BANK1"
.segment "BANK2"
.segment "BANK3"
.segment "BANK4"
.segment "BANK5"
.segment "BANK6"
.segment "BANK7"
.segment "BANK8"
.segment "BANK9"
.segment "BANK10"
.segment "BANK11"
.segment "BANK12"


.segment "STARTUP"
; this should be mapped to the last PRG bank

start:
_exit:

    sei
	cld
	ldx #$40
	stx CTRL_PORT2
    ldx #$ff
    txs
    inx
    stx PPU_MASK
    stx DMC_FREQ
    stx PPU_CTRL		;no NMI
	
	jsr _disable_irq ;disable mmc3 IRQ
	
	;x is zero

initPPU:
    bit PPU_STATUS
@1:
    bit PPU_STATUS
    bpl @1
@2:
    bit PPU_STATUS
    bpl @2

clearPalette:
	lda #$3f
	sta PPU_ADDR
	stx PPU_ADDR
	lda #$0f
	ldx #$20
@1:
	sta PPU_DATA
	dex
	bne @1
	
	
;	lda #$01				; DEBUGGING
;	sta PPU_DATA
;	lda #$11
;	sta PPU_DATA
;	lda #$21
;	sta PPU_DATA
;	lda #$30
;	sta PPU_DATA

clearVRAM:
	txa
	ldy #$20
	sty PPU_ADDR
	sta PPU_ADDR
	ldy #$10
@1:
	sta PPU_DATA
	inx
	bne @1
	dey
	bne @1

clearRAM:
    txa
@1:
    sta $000,x
    sta $100,x
    sta $200,x
    sta $300,x
    sta $400,x
    sta $500,x
    sta $600,x
    sta $700,x
    inx
    bne @1
	
; don't call any subroutines until the banks are in place	
	
	
	
; MMC3 reset

; set which bank at $8000
; also $c000 fixed to 14 of 15
	lda #0 ; PRG bank zero
	jsr _set_prg_8000
; set which bank at $a000
	lda #13 ; PRG bank 13 of 15
	jsr _set_prg_a000
	
; with CHR invert, set $0000-$03FF
	lda #0
	jsr _set_chr_mode_2
; with CHR invert, set $0400-$07FF
	lda #1
	jsr _set_chr_mode_3
; with CHR invert, set $0800-$0BFF
	lda #2
	jsr _set_chr_mode_4
; with CHR invert, set $0C00-$0FFF
	lda #3
	jsr _set_chr_mode_5
; with CHR invert, set $1000-$17FF
	lda #4
	jsr _set_chr_mode_0
; with CHR invert, set $1800-$1FFF
	lda #6
	jsr _set_chr_mode_1
;set mirroring to vertical, no good reason	
	lda #0
	jsr _set_mirroring
;allow reads and writes to WRAM	
	lda #$80 ;WRAM_ON 0x80
	jsr _set_wram_mode
	
	cli ;allow irq's to happen on the 6502 chip	
		;however, the mmc3 IRQ was disabled above
	
	

	lda #4
	jsr _pal_bright
	jsr _pal_clear
	jsr _oam_clear

    jsr zerobss
	jsr	copydata

    lda #<(__STACK_START__+__STACKSIZE__) ;changed
    sta	sp
    lda	#>(__STACK_START__+__STACKSIZE__)
    sta	sp+1            ; Set argument stack ptr

;	jsr	initlib
; removed. this called the CONDES function





	








	lda #%10000000
	sta <PPU_CTRL_VAR
	sta PPU_CTRL		;enable NMI
	lda #%00000110
	sta <PPU_MASK_VAR

waitSync3:
	lda <FRAME_CNT1
@1:
	cmp <FRAME_CNT1
	beq @1

detectNTSC:
	ldx #52				;blargg's code
	ldy #24
@1:
	dex
	bne @1
	dey
	bne @1

	lda PPU_STATUS
	and #$80
	sta <NTSC_MODE

	jsr _ppu_off

	lda #0
	ldx #0
	jsr _set_vram_update

	lda #$fd
	sta <RAND_SEED
	sta <RAND_SEED+1

	lda #0
	sta PPU_SCROLL
	sta PPU_SCROLL
	
	
	
;BANK12
	
	lda #SOUND_BANK ;swap the music in place before using
					;SOUND_BANK is defined above
	jsr _set_prg_8000
	
	ldx #<music_data
	ldy #>music_data
	lda <NTSC_MODE
	jsr FamiToneInit

	ldx #<sounds_data
	ldy #>sounds_data
	jsr FamiToneSfxInit
	
	lda #$00 ;PRG bank #0 at $8000, back to basic
	jsr _set_prg_8000

	jmp _main			;no parameters
	
	

	.include "MMC3/mmc3_code.asm"
	.include "LIB/neslib.s"
	.include "LIB/nesdoug.s"
	



	

	
.segment "CODE"	
	.include "MUSIC/famitone2.s"
; When music files get very big, it's probably best to
; split the songs into multiple swapped banks
; the music code itself is in the regular CODE banks.
; It could be moved into BANK12 if music data is small.
	
.segment "BANK12"	
	
music_data:
	.include "MUSIC/TestMusic3.s"

sounds_data:
	.include "MUSIC/SoundFx.s"


	
	
;.segment "SAMPLES"
;	.incbin "MUSIC/BassDrum.dmc"



.segment "VECTORS"

    .word nmi	;$fffa vblank nmi
    .word start	;$fffc reset
   	.word irq	;$fffe irq / brk


.segment "CHARS"

	.incbin "Alpha.chr"
	.incbin "Gears.chr"
; the CHARS segment is much bigger, and I could have 
; incbin-ed many more chr files
	


hello_mmc3.c
/*	example of MMC3 for cc65
 *	Doug Fraker 2019
 *	Feb 2023 version
 */	
 
#include "LIB/neslib.h"
#include "LIB/nesdoug.h" 
#include "MMC3/mmc3_code.h"
#include "MMC3/mmc3_code.c"
#include "Sprites.h"

 
#pragma bss-name(push, "ZEROPAGE")

// GLOBAL VARIABLES

unsigned char arg1;
unsigned char arg2;
unsigned char pad1;
unsigned char pad1_new;
unsigned char char_state;

// fixed point 8.8
unsigned int scroll_top;
unsigned int scroll2;
unsigned int scroll3;
unsigned int scroll4;
unsigned char temp;

unsigned char sprite_x;
unsigned char sprite_y;
unsigned char dirLR;


#pragma bss-name(pop)
// should be in the regular 0x300 ram now

unsigned char irq_array[32];
unsigned char double_buffer[32];


#pragma bss-name(push, "XRAM")
// extra RAM at $6000-$7fff
unsigned char wram_array[0x2000];

#pragma bss-name(pop)






const unsigned char palette_bg[]={
0x0f, 0, 0x10, 0x30,
0x0f, 0, 0, 0,
0x0f, 0, 0, 0,
0x0f, 0, 0, 0
}; 

const unsigned char palette_spr[]={
0x0f, 0x09, 0x19, 0x29, // greens
0x0f, 0, 0, 0,
0x0f, 0, 0, 0,
0x0f, 0, 0, 0
}; 


// test putting things in other banks
#pragma rodata-name ("BANK0")
#pragma code-name ("BANK0")
const unsigned char TEXT0[]="BANK0";

void function_bank0(void){
	ppu_off();
	vram_adr(NTADR_A(1,4));
	vram_write(TEXT0,sizeof(TEXT0));
	ppu_on_all();
}


#pragma rodata-name ("BANK1")
#pragma code-name ("BANK1")
const unsigned char TEXT1[]="BANK1";

void function_bank2(void);	// a prototype can be anywhere
							// as long as it's above the code that uses it

void function_bank1(void){
	ppu_off();
	vram_adr(NTADR_A(1,6));
	vram_write(TEXT1,sizeof(TEXT1));
	ppu_on_all();
	
	banked_call(2, function_bank2);
	// calling a function in another bank, use banked_call()
}


#pragma rodata-name ("BANK2")
#pragma code-name ("BANK2")
const unsigned char TEXT2[]="BANK2";

void function_same_bank(){
	vram_put(0);
	vram_put('H');
	vram_put('I');
}

void function_bank2(void){
	ppu_off();
	vram_adr(NTADR_A(1,8));
	vram_write(TEXT2,sizeof(TEXT2));

	function_same_bank();
	// calling a function in same bank, use regular function calls
	
	ppu_on_all();
}


#pragma rodata-name ("BANK3")
#pragma code-name ("BANK3")
const unsigned char TEXT3[]="BANK3";

void function_bank3(void){
	 
	ppu_off();
	vram_adr(NTADR_A(1,10));
	vram_write(TEXT3,sizeof(TEXT3));
	
	vram_put(0);
	vram_put(arg1); // these args were passed via globals
	vram_put(arg2);
	ppu_on_all();
}


#pragma rodata-name ("BANK6")
#pragma code-name ("BANK6")
const unsigned char TEXT6[]="BANK6";

void function_bank6(void){
	 
	ppu_off();
	vram_adr(NTADR_A(1,14));
	vram_write(TEXT6,sizeof(TEXT6));
	
	vram_put(0);
	vram_put(wram_array[0]); // testing the $6000-7fff area
	vram_put(wram_array[2]); // should print A, C
	ppu_on_all();
}
	
	
	
	
// the fixed bank, bank 7
	
#pragma rodata-name ("CODE")
#pragma code-name ("CODE")	

const unsigned char text[]="BACK IN FIXED BANK";

void draw_sprites (void);




void main (void) {
	
	set_mirroring(MIRROR_HORIZONTAL);
	bank_spr(1);
	irq_array[0] = 0xff; // end of data
	set_irq_ptr(irq_array); // point to this array
	
	
	// clear the WRAM, not done by the init code
	// memfill(void *dst,unsigned char value,unsigned int len);
	memfill(wram_array,0,0x2000); 
	
	
	wram_array[0] = 'A'; // put some values at $6000-7fff
	wram_array[2] = 'C'; // for later testing

	
	ppu_off(); // screen off
	pal_bg(palette_bg); //	load the BG palette
	pal_spr(palette_spr); // load the sprite palette
	
	// draw some things
	vram_adr(NTADR_A(20,3)); // gear and squares
	vram_put(0xc0);
	vram_put(0xc1);
	vram_put(0xc2);
	vram_put(0xc3);
	vram_adr(NTADR_A(20,4));
	vram_put(0xd0);
	vram_put(0xd1);
	vram_put(0xd2);
	vram_put(0xd3);
	vram_adr(NTADR_A(20,7));
	vram_put(0xc0);
	vram_put(0xc1);
	vram_put(0xc2);
	vram_put(0xc3);
	vram_adr(NTADR_A(20,8));
	vram_put(0xd0);
	vram_put(0xd1);
	vram_put(0xd2);
	vram_put(0xd3);
	
	vram_adr(NTADR_A(20,5)); // blocks of color
	vram_put(0x2);
	vram_put(0x2);
	vram_put(0x2);
	vram_put(0x2);
	vram_adr(NTADR_A(20,9));
	vram_put(0x2);
	vram_put(0x2);
	vram_put(0x2);
	vram_put(0x2);
	vram_adr(NTADR_A(20,13));
	vram_put(0x2);
	vram_put(0x2);
	vram_put(0x2);
	vram_put(0x2);
    
	music_play(0);
	//ppu_on_all(); // turn on screen
	
	set_chr_mode_5(8); // make sure the gear tiles loaded
                       // for the first few frames
	
	
	// calling functions in other banks
	
	banked_call(0, function_bank0);
	banked_call(1, function_bank1);
	// banked_call(2, function_bank2); // moved to bank 1
	
	arg1 = 'G'; // must pass arguments with globals
	arg2 = '4';
	banked_call(3, function_bank3);
	banked_call(6, function_bank6);

	 
	ppu_off(); // screen off
	vram_adr(NTADR_A(1,16));
	vram_write(text,sizeof(text));

	sprite_x = 0x50;
	sprite_y = 0x30;
	draw_sprites();
	
	ppu_on_all(); //	turn on screen
	

	
	while (1){ // infinite loop
		ppu_wait_nmi();
		
		pad1 = pad_poll(0);
		pad1_new = get_pad_new(0);
		
		if(pad1 & PAD_A){ // shift screen right = subtract from scroll
			scroll_top -= 0x80; // sub pixel movement
			scroll2 -= 0x100; // 1 pixel
			scroll3 -= 0x180;
			scroll4 -= 0x200; 
		}
		
		if(pad1 & PAD_B){ // shift screen right = subtract from scroll
			scroll_top += 0x80; // sub pixel movement
			scroll2 += 0x100; // 1 pixel
			scroll3 += 0x180;
			scroll4 += 0x200;
		}
		
		temp = scroll_top >> 8;
		set_scroll_x(temp);
		
		if((get_frame_count() & 0x03) == 0){ // every 4th frame
			++char_state;
			if(char_state >=4) char_state = 0;
		}
		
		
		// load the irq array with values it can parse
		// ! CHANGED it, double buffered so we aren't editing the same
		// array that the irq system is reading from
		
		// top of the screen, probably read in v-blank
		// put whole screen things here, before the split
		double_buffer[0] = 0xfc; // CHR mode 5, change the 0xc00-0xfff tiles
		double_buffer[1] = 8; // top of the screen, static value
		double_buffer[2] = 47; // value < 0xf0 = scanline count, 1 less than #

		// after the first split
		double_buffer[3] = 0xf5; // H scroll change, do first for timing
		temp = scroll2 >> 8;
		double_buffer[4] = temp; // scroll value
		double_buffer[5] = 0xfc; // CHR mode 5, change the 0xc00-0xfff tiles
		double_buffer[6] = 8 + char_state; // value = 8,9,10,or 11
		double_buffer[7] = 30; // scanline count
		
		// after the 2nd split
		double_buffer[8] = 0xf5; // H scroll change
		temp = scroll3 >> 8;
		double_buffer[9] = temp; // scroll value
			double_buffer[10] = 0xf1; // $2001 test changing color emphasis
			double_buffer[11] = 0xfe; // value COL_EMP_DARK 0xe0 + 0x1e
			//double_buffer[11] = 0x1f; // alternate value for grayscale
		double_buffer[12] = 30; // scanline count
		
		// after the 3rd split
		double_buffer[13] = 0xf5; // H scroll change
		temp = scroll4 >> 8;
		double_buffer[14] = temp; // scroll value
		double_buffer[15] = 30; // scanline count
		
		double_buffer[16] = 0xf5; // H scroll change
		double_buffer[17] = 0; // to zero, to set the fine X scroll
		double_buffer[18] = 0xf6; // 2 writes to 2006 shifts screen
		double_buffer[19] = 0x20; // need 2 values...
		double_buffer[20] = 0x00; // PPU address $2000 = top of screen
		
		double_buffer[21] = 0xff; // end of data
		
		
		if(pad1 & PAD_LEFT){ // shift screen right = subtract from scroll
			sprite_x -= 1;
			dirLR = 0;
		}
		else if(pad1 & PAD_RIGHT){ // shift screen right = subtract from scroll
			sprite_x += 1;
			dirLR = 1;
		}
		if(pad1 & PAD_UP){ // shift screen right = subtract from scroll
			sprite_y -= 1;
		}
		else if(pad1 & PAD_DOWN){ // shift screen right = subtract from scroll
			sprite_y += 1;
		}
		
		draw_sprites();
		
		// wait till the irq system is done before changing it
		// this could waste a lot of CPU time, so we do it last
		while(!is_irq_done() ){ // have we reached the 0xff, end of data
							// is_irq_done() returns zero if not done
			// do nothing while we wait
		}
		
		// copy from double_buffer to the irq_array
		// memcpy(void *dst,void *src,unsigned int len);
		memcpy(irq_array, double_buffer, sizeof(irq_array)); 
	}
}



	
// testing sprites using the second tileset	
void draw_sprites (void) { 
	oam_clear();
	if(!dirLR) { // left
		oam_meta_spr(sprite_x, sprite_y, RoundSprL);
	}
	else {
		oam_meta_spr(sprite_x, sprite_y, RoundSprR);
	}
	
}



MMC3/mmc3_code.asm
;version Feb 2023
;credit to Matt Hughson for help


.define A12_INVERT $80
;values 0 or $80, for the $8000 register
;(changes in tileset mapping)

;if invert bit is 0 
;mode 0 changes $0000-$07FF
;mode 1 changes $0800-$0FFF
;mode 2 changes $1000-$13FF
;mode 3 changes $1400-$17FF
;mode 4 changes $1800-$1BFF
;mode 5 changes $1C00-$1FFF

;if invert bit is $80
;mode 0 changes $1000-$17FF
;mode 1 changes $1800-$1FFF
;mode 2 changes $0000-$03FF
;mode 3 changes $0400-$07FF
;mode 4 changes $0800-$0BFF
;mode 5 changes $0C00-$0FFF



.segment "ZEROPAGE"
    BP_BANK_8000: .res 1
;   mmc3_8000_PRG:    .res 1
;    mmc3_8001_PRG:    .res 1
;    mmc3_8000_CHR:    .res 1
;    mmc3_8001_CHR:    .res 1
    mmc3_8000: .res 1 ; backup, just in case interrupt
    mmc3_ptr:        .res 2 ; array for the irq parser
    mmc3_index:        .res 1 ; index to this array
    irq_done:        .res 1

    
;   .exportzp mmc3_8000_PRG, mmc3_8001_PRG, mmc3_8000_CHR, mmc3_8001_CHR
    .exportzp BP_BANK_8000, mmc3_8000
    .exportzp mmc3_ptr, mmc3_index, irq_done
    

.segment "STARTUP"
;needs to be mapped to the fixed bank

.export _set_prg_8000, _get_prg_8000, _set_prg_a000
.export _set_chr_mode_0, _set_chr_mode_1, _set_chr_mode_2, _set_chr_mode_3
.export _set_chr_mode_4, _set_chr_mode_5

.export _set_mirroring, _set_wram_mode, _disable_irq
.export _set_irq_ptr, _is_irq_done


; most of these functions only use the A register
; because some of the music code needs X unchanged


; sets the bank at $8000-9fff
_set_prg_8000:
    sta BP_BANK_8000
    pha
    lda #(6 | A12_INVERT)
    bne bank_swap_common

;    sta mmc3_8000 ;backup, in case of interrupt
;    sta $8000
;    pla
;    sta $8001
;    rts


; returns the current bank at $8000-9fff    
_get_prg_8000:
    lda BP_BANK_8000
    ldx #0
    rts
    
    
; sets the bank at $a000-bfff
_set_prg_a000:
    pha
    lda #(7 | A12_INVERT)
    bne bank_swap_common
    
;    sta mmc3_8000 ;backup, in case of interrupt
;    sta $8000
;    pla
;    sta $8001
;    rts



; see chart in above for what these do
_set_chr_mode_0:
    pha
    lda #(0 | A12_INVERT)
bank_swap_common:    
    sta mmc3_8000 ;backup, in case of interrupt
    sta $8000   
    pla
    sta $8001
    rts

    
_set_chr_mode_1:
    pha
    lda #(1 | A12_INVERT)
    bne bank_swap_common
    
;    sta mmc3_8000 ;backup, in case of interrupt
;    sta $8000   
;    pla
;    sta $8001
;    rts

    
_set_chr_mode_2:
    pha
    lda #(2 | A12_INVERT)
    bne bank_swap_common
    
;    sta mmc3_8000 ;backup, in case of interrupt
;    sta $8000   
;    pla
;    sta $8001
;    rts

    
_set_chr_mode_3:
    pha
    lda #(3 | A12_INVERT)
    bne bank_swap_common
    
;    sta mmc3_8000 ;backup, in case of interrupt
;    sta $8000   
;    pla
;    sta $8001
;    rts

    
_set_chr_mode_4:
    pha
    lda #(4 | A12_INVERT)
    bne bank_swap_common
    
;    sta mmc3_8000 ;backup, in case of interrupt
;    sta $8000   
;    pla
;    sta $8001
;    rts

    
_set_chr_mode_5:
    pha
    lda #(5 | A12_INVERT)
    bne bank_swap_common
    
;    sta mmc3_8000 ;backup, in case of interrupt
;    sta $8000   
;    pla
;    sta $8001
;    rts

 
    
; MIRROR_VERTICAL 0
; MIRROR_HORIZONTAL 1    
_set_mirroring:
    sta $a000
    rts
    
    
; WRAM_OFF $40
; WRAM_ON $80
; WRAM_READ_ONLY $C0
_set_wram_mode:
    sta $a001
    rts
    

_disable_irq:
    sta $e000 ;any value
    lda #<default_array
    ldx #>default_array
    ;jmp _set_irq_ptr ; fall through
    
    
_set_irq_ptr:
;ax = pointer
    sta mmc3_ptr
    stx mmc3_ptr+1
    rts    
    
    
_is_irq_done:
    lda irq_done
    ldx #0
    rts

    
default_array: ;just an eof terminator
.byte $ff



    
irq:
    pha
    txa
    pha
    tya
    pha
    
    sta $e000    ; disable mmc3 irq
                ; any value will do
    
    jsr irq_parser
    
    pla
    tay
    pla
    tax
    pla
    rti
    
    
;format
;value < 0xf0, it's a scanline count
;zero is valid, it triggers an IRQ at the end of the current line

;if >= 0xf0...
;f0 = 2000 write, next byte is write value
;f1 = 2001 write, next byte is write value
;f2-f4 unused - future TODO ?
;f5 = 2005 write, next byte is H Scroll value
;f6 = 2006 write, next 2 bytes are write values


;f7 = change CHR mode 0, next byte is write value
;f8 = change CHR mode 1, next byte is write value
;f9 = change CHR mode 2, next byte is write value
;fa = change CHR mode 3, next byte is write value
;fb = change CHR mode 4, next byte is write value
;fc = change CHR mode 5, next byte is write value

;fd = very short wait, no following byte 
;fe = short wait, next byte is quick loop value
;(for fine tuning timing of things)

;ff = end of data set

    
irq_parser:
    ldy mmc3_index
;    ldx #0
@loop:
    lda (mmc3_ptr), y ; get value from array
    iny
    cmp #$fd ;very short wait
    beq @loop
    
    cmp #$fe ;fe-ff wait or exit
    bcs @wait
    
    cmp #$f0
    bcs @1
    jmp @scanline ;below f0
@1:    
    
    cmp #$f7
    bcs @chr_change
;f0-f6    
    tax
    lda (mmc3_ptr), y ; get value from array
    iny
    cpx #$f0
    bne @2
    sta $2000 ; f0
    jmp @loop
@2:
    cpx #$f1
    bne @3
    sta $2001 ; f1
    jmp @loop
@3:
    cpx #$f5 
    bne @4
    ldx #4
@better_timing: ; don't change till near the end of the line
    dex
    bne @better_timing
    
    sta $2005 ; f5
    sta $2005 ; second value doesn't matter
    jmp @loop
@4:
    sta $2006 ; f6
    lda (mmc3_ptr), y ; get 2nd value from array
    iny    
    sta $2006
    jmp @loop
    
@wait: ; fe-ff wait or exit
    cmp #$ff
    beq @exit    
    lda (mmc3_ptr), y ; get value from array
    iny
    tax
    beq @loop ; if zero, just exit
@wait_loop: ; the timing of this wait could change if this crosses a page boundary
    dex
    bne @wait_loop        
    jmp @loop    

@chr_change:
;f7-fc change a CHR set
    sec
    sbc #$f7 ;should result in 0-5
    ora #A12_INVERT
    sta $8000
    lda (mmc3_ptr), y ; get next value
    iny
    sta $8001

    lda mmc3_8000 ;restore the MMC3 bank select register
    sta $8000     ;in case we interrupted something
    jmp @loop
    
@scanline:
    nop ;trying to improve stability
    nop
    nop
    nop
    jsr set_scanline_count ;this terminates the set
    sty mmc3_index
    rts
    
@exit:
    sta irq_done ;value 0xff
    dey ; undo the previous iny, keep it pointed to ff
    sty mmc3_index
    rts
    
    
set_scanline_count:
; any value will do for most of these registers
    sta $e000 ; disable mmc3 irq
    sta $c000 ; set the irq counter reload value
    sta $c001 ; reload the reload value
    sta $e001 ; enable the mmc3 irq
    cli ;make sure irqs are enabled
    rts


    

    


    
    
    